The last of the getTimeFieldSpec calls

npawar opened the pull request:
https://github.com/apache/incubator-pinot/issues/2756
Removing the last of the getTimeFieldSpec calls.

1) SegmentGeneratorConfig now uses the new getSpecForTimeColumn() call, which returns a DateTimeFieldSpec. T**ime column is expected in table config if it is to be considered when creating segment**. (this also sets the stage for allowing a DateTimeFieldSpec to be a primary time column)
2) Removing TimeFieldSpec special handling from RealtimeSegmentConverter.getUpdatedSchema. There is no need to remove the incoming time spec. Neither the recordReader nor the dataSource require the updated schema. Plus, the record transformer is a pass through.
3) Removing special casing for TimeFieldSpec in Schema.isBackwardCompatible() method. The for loop for all specs includes time

Jackie-Jiang commented on pinot-core/src/main/java/org/apache/pinot/core/realtime/converter/RealtimeSegmentConverter.java:
```diff
@@ -156,20 +156,8 @@ public void build(@Nullable SegmentVersion segmentVersion, ServerMetrics serverM
   @VisibleForTesting
   public Schema getUpdatedSchema(Schema original) {
```
We need to remove the time transform here because the time has already been transformed.
Please also update the javadoc

npawar replied:
I believe we don't need to change the schema because:
1) The RecordTransformer is a pass through
2) The RecordReader is just looping over physical column names so it only has outgoing

We shouldn't need to remove incoming, because that implies that we need to remove transform functions from all field specs.

Jackie-Jiang commented on pinot-core/src/main/java/org/apache/pinot/core/indexsegment/generator/SegmentGeneratorConfig.java:
```diff
@@ -182,42 +180,21 @@ public SegmentGeneratorConfig(TableConfig tableConfig, Schema schema) {
-   * Set time column details using the given time column. If not found, use schema
+   * Set time column details using the given time column
    */
   public void setTime(String timeColumnName, Schema schema) {
```
Annotate `timeColumnName` as nullable or check it on the caller side

Jackie-Jiang commented on pinot-core/src/test/java/org/apache/pinot/core/indexsegment/generator/SegmentGeneratorConfigTest.java:
```diff
@@ -48,14 +48,12 @@ public void testEpochTime() {
     assertNull(segmentGeneratorConfig.getSimpleDateFormat());
 
     // table config not provided
```
Remove this line?

Jackie-Jiang commented on pinot-core/src/test/java/org/apache/pinot/core/indexsegment/generator/SegmentGeneratorConfigTest.java:
```diff
@@ -74,14 +72,12 @@ public void testSimpleDateFormat() {
     assertEquals(segmentGeneratorConfig.getSimpleDateFormat(), "yyyyMMdd");
 
     // Table config not provided
```
Remove this line?

Jackie-Jiang commented on pinot-core/src/test/java/org/apache/pinot/realtime/converter/RealtimeSegmentConverterTest.java:
```diff
@@ -46,28 +46,11 @@ public void testNoVirtualColumnsInSchema() {
-
-  @Test
-  public void testNoTimeColumnsInSchema() {
-    Schema schema = new Schema();
```
Do we still want this support? (Realtime table without time column)

npawar replied:
realtime table without time will never happen right?
Also after changing the getUpdatedSchema method, this test becomes irrelevant

Jackie-Jiang replied:
I think @fx19880617 added that support. Any opinion on this?

Jackie-Jiang commented on pinot-spi/src/main/java/org/apache/pinot/spi/data/Schema.java:
```diff
@@ -582,10 +582,6 @@ public boolean equals(Object o) {
-      return false;
-    }
     for (Map.Entry<String, FieldSpec> entry : oldSchema.getFieldSpecMap().entrySet()) {
       if (!getColumnNames().contains(entry.getKey())) {
```
Move `getColumnNames()` out of the loop

npawar commented:
> Not related to this PR, but inside the Schema can we do the TimeFieldSpec to DateTimeFieldSpec conversion on the setter side instead of the getter side?
> The problem of conversion on the getter side is that the schema still stores the deprecated TimeFieldSpec, and `getFieldSpecMap()` and `getFieldSpecFor()` will still return the TimeFieldSpec instead of DateTimeFieldSpec. This can easily cause bugs inside the code.

That was exactly the initial plan - to modify the setter to save it as a DateTimeFieldspec. But now we changed the plan. We are not going to do any changes to ser/deser. Reason mainly being external integrations (presto, superset etc) which still need TimeFieldSpec, and also upgrades. I've updated the design doc: https://docs.google.com/document/d/1SU1jCjfsIDSA960fD5YWQbD72p8UdGF0c7CroFNt9Ho/edit#heading=h.6gaudq1t6ys8 (Step C and also put the previous approach in Discarded section)

New plan is to make sure Pinot can treat TIME or DATE_TIME either of them as the primary time column.
